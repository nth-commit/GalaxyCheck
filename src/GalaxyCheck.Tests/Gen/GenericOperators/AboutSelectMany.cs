using FsCheck.Xunit;
using Xunit;
using System;
using GalaxyCheck;
using GC = GalaxyCheck;
using static Tests.TestUtils;

namespace Tests.Gen.GenericOperators
{
    [Properties(Arbitrary = new [] { typeof(ArbitraryGen) }, MaxTest = 10)]
    public class AboutSelectMany
    {
        [Fact]
        public void Snapshots()
        {
            var gen0 = GC.Gen.Int32().Between(0, 5);
            SnapshotGenExampleSpaces(
                from x in gen0
                from y in gen0
                select (x, y));
        }

        [Property]
        public void ItExhaustsWhenLeftGenHasImpossiblePredicate(IGen<object> leftGen, IGen<object> rightGen)
        {
            var gen =
                from x in leftGen
                where false
                from y in rightGen
                select (x, y);

            TestWithSeed(seed =>
            {
                Action test = () => gen.Sample(seed: seed);

                Assert.Throws<GC.Exceptions.GenExhaustionException>(test);
            });
        }

        [Property]
        public void ItExhaustsWhenRightGenHasImpossiblePredicate(IGen<object> leftGen, IGen<object> rightGen)
        {
            var gen =
                from x in leftGen
                from y in rightGen
                where false
                select (x, y);

            TestWithSeed(seed =>
            {
                Action test = () => gen.Sample(seed: seed);

                Assert.Throws<GC.Exceptions.GenExhaustionException>(test);
            });
        }
    }
}
