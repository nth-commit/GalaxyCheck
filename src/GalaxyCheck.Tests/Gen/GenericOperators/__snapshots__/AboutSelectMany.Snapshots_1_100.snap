(1, 5)
..(0, 5)
....(0, 0)
....(0, 3)
......(0, 2)
........(0, 1)
....(0, 4)
......(0, 2)
........(0, 1)
..(1, 0)
..(1, 3)
....(1, 2)
......(1, 1)
..(1, 4)
....(1, 2)
......(1, 1)
